contracts/cache_backend.go

```diff
@@ -0,0 +1,114 @@
+package contracts
+
+import (
+	"math/big"
+
+	"github.com/celo-org/celo-blockchain/common"
+	"github.com/celo-org/celo-blockchain/core/vm"
+	"github.com/celo-org/celo-blockchain/metrics"
+	lru "github.com/hashicorp/golang-lru"
+)
+
+var (
+	cacheHits    = metrics.NewRegisteredMeter("contract_comm/caller/cache_hits", nil)
+	cacheMisses  = metrics.NewRegisteredMeter("contract_comm/caller/cache_misses", nil)
+	cacheBadCast = metrics.NewRegisteredMeter("contract_comm/caller/cache_bad_cast", nil)
+)
+
+type cacheKey struct {
+	stateRoot common.Hash
+	caller    vm.ContractRef
+	addr      common.Address
+	input     []byte
+	gas       uint64
+}
+
+type cacheResult struct {
+	ret     []byte
+	gasLeft uint64
+}
+
+type cacheBackend struct {
+	source Backend
+	cache  *lru.Cache
+}
+
+func WithCache(backend Backend, cache *lru.Cache) Backend {
+	return &cacheBackend{
+		source: backend,
+		cache:  cache,
+	}
+}
+
+func createGlobalCache() func(Backend) Backend {
+	var globalCache, err = lru.New(100)
+	if err != nil {
+		panic("error creating global cache: " + err.Error())
+	}
+	return func(backend Backend) Backend {
+		return WithCache(backend, globalCache)
+	}
+}
+
+var WithGlobalCache = createGlobalCache()
```

Comment:
I don't like the name of `createGlobalCache()`, since it suggests it returns the cache rather than a function to wrap a backend with the cache.  And, really, what makes it global isn't anything in the function itself but rather how it's used on this line to set a global var.  We could inline it, but the result isn't super-readable, so maybe just rename it to something ugly like `createWrapperWithGlobalCache()`or something ugly like that.